### PR TITLE
[enhancement](nereids) Support or to in conversion

### PR DESCRIPTION
**Pull request by Kikyou1997**
## Proposed changes

Add new rule named "OrToIn", used to convert multi equalTo which has same slot and compare to a literal of disjunction to a InPredicate so that it could be pushdown to storage engine.

for example:

```sql
col1 = 1 or col1 = 2 or col1 = 3 and (col2 = 4)
col1 = 1 and col1 = 3 and col2 = 3 or col2 = 4
(col1 = 1 or col1 = 2) and  (col2 = 3 or col2 = 4)
```

would be converted to 

```sql
col1 in (1, 2) or col1 = 3 and (col2 = 4)
col1 = 1 and col1 = 3 and col2 = 3 or col2 = 4
(col1 in (1, 2) and (col2 in (3, 4)))
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/ExpressionRewriteContext.java [hunk 1]
```diff
@@ -18,14 +18,33 @@
...
+import java.util.Map;
+import java.util.Set;
 
 /**
  * expression rewrite context.
  */
 public class ExpressionRewriteContext {
+
     public final CascadesContext cascadesContext;
+    public Map<NamedExpression, Set<Literal>> slotNameToLiteral = new HashMap<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/ExpressionRewriteContext.java:36]: not add rule sepcific attr to this context

File: fe/fe-core/src/main/java/org/apache/doris/common/util/NetUtils.java [hunk 2]
```diff
@@ -100,30 +97,7 @@ public static String getIpByHost(String host) throws UnknownHostException {
...
-            if (ss != null) {
-                try {
-                    ss.close();
-                } catch (IOException e) {
-                    /* should not be thrown */
-                }
-            }
-        }
-        return false;
+        return true;
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/NetUtils.java:100]: why always return true?

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/common/util/NetUtils.java:100]: This is a stupid mistake, sorry.

**Comment by Kikyou1997** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OrToIn.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ * (col1 = 1 or col1 = 2) and  (col2 = 3 or col2 = 4)
+ * <p>
+ * would be converted to:
+ * col1 in (1, 2) or col1 = 3 and (col2 = 4)
+ * col1 = 1 and col1 = 3 and col2 = 3 or col2 = 4
+ * (col1 in (1, 2) and (col2 in (3, 4)))
+ */
+public class OrToIn extends AbstractExpressionRewriteRule {
+
+    private static final int CONVERT_TO_IN_THRESHOLD = 2;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/OrToIn.java:56]: use session variable REWRITE_OR_TO_IN_PREDICATE_THRESHOLD

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Comment by englefly** [comment]
this pr does not cover some cases which are covered by original planner.
refer to ut: testWideCommonFactorsWithEqualPredicate()
(tb1.k1=1 and tb2.k1=1) or (tb1.k1 =2 and tb2.k1=2) => tb1.k1 in (1, 2) and tb2.k1 in (1, 2) and [...]
we need another pr to fix this.

**Review by englefly — APPROVED** [review]

**Pull request merged**